arangod/Aql/SortExecutor.cpp

```diff
@@ -135,146 +134,129 @@ SortExecutor::~SortExecutor() {
+
+  if (!_inputReady) {
+    consumeInput(inputRange, state);
+    if (inputRange.upstreamState() == ExecutorState::HASMORE) {
+      return {state, NoStats{}, upstreamCall};
+    }
+    if (_returnNext < _rowIndexes.size()) {
+      doSorting();
+      _inputReady = true;
+    }
+  }
+
+  while (_returnNext < _rowIndexes.size() && !output.isFull()) {
+    InputAqlItemRow inRow(_inputBlocks[_rowIndexes[_returnNext].first],
+                          _rowIndexes[_returnNext].second);
     output.copyRow(inRow);
     output.advanceRow();
     _returnNext++;
   }
 
-  if (_returnNext >= _sortedIndexes.size()) {
-    return {ExecutorState::DONE, NoStats{}, upstreamCall};
+  if (_returnNext >= _rowIndexes.size()) {
+    state = ExecutorState::DONE;
+  } else {
+    state = ExecutorState::HASMORE;
   }
-  return {ExecutorState::HASMORE, NoStats{}, upstreamCall};
+
+  return {state, NoStats{}, upstreamCall};
 }
 
 void SortExecutor::doSorting() {
   TRI_IF_FAILURE("SortBlock::doSorting") {
     THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
   }
 
-  size_t memoryUsageForRowIndexes = _input->memoryUsageForRowIndexes();
-  // may throw
-  ResourceUsageScope guard(_infos.getResourceMonitor(),
-                           memoryUsageForRowIndexes);
-
-  TRI_ASSERT(_input != nullptr);
-  _sortedIndexes = _input->produceRowIndexes();
-
-  // now we are responsible for tracking the memory
-  guard.steal();
-  _memoryUsageForRowIndexes = memoryUsageForRowIndexes;
-
   // comparison function
-  OurLessThan ourLessThan(_infos.vpackOptions(), *_input,
+  OurLessThan ourLessThan(_infos.vpackOptions(), _inputBlocks,
                           _infos.sortRegisters());
   if (_infos.stable()) {
-    std::stable_sort(_sortedIndexes.begin(), _sortedIndexes.end(), ourLessThan);
+    std::stable_sort(_rowIndexes.begin(), _rowIndexes.end(), ourLessThan);
   } else {
-    std::sort(_sortedIndexes.begin(), _sortedIndexes.end(), ourLessThan);
+    std::sort(_rowIndexes.begin(), _rowIndexes.end(), ourLessThan);
   }
 }
 
 std::tuple<ExecutorState, NoStats, size_t, AqlCall> SortExecutor::skipRowsRange(
-    AqlItemBlockInputMatrix& inputMatrix, AqlCall& call) {
+    AqlItemBlockInputRange& inputRange, AqlCall& call) {
   AqlCall upstreamCall{};
 
-  if (inputMatrix.upstreamState() == ExecutorState::HASMORE) {
-    // If our inputMatrix does not contain all upstream rows
-    return {ExecutorState::HASMORE, NoStats{}, 0, upstreamCall};
-  }
+  ExecutorState state = ExecutorState::HASMORE;
 
-  if (_input == nullptr) {
-    initializeInputMatrix(inputMatrix);
+  if (!_inputReady) {
+    auto inputBlock = inputRange.getBlock();
+    if (inputBlock != nullptr) {
+      _inputBlocks.emplace_back(inputBlock);
+    }
   }
 
-  if (_returnNext >= _sortedIndexes.size()) {
+  if (_returnNext >= _rowIndexes.size() && !_rowIndexes.empty()) {
     // Bail out if called too often,
     // Bail out on no elements
     return {ExecutorState::DONE, NoStats{}, 0, upstreamCall};
   }
```

Comment:
I don't think this entire block is necessary. 
If `_inputReady` is false, it has no effect, because `_rowIndexes` will still be empty.
If `_inputReady` is true, we don't need it, because the following `if`s and `while`s take care of the situation already.

Reply:
Addressed